src/main/java/uk/gov/hmcts/reform/slc/services/servicebus/MessageProcessor.java

```diff
@@ -0,0 +1,70 @@
+package uk.gov.hmcts.reform.slc.services.servicebus;
+
+import com.microsoft.azure.servicebus.IMessage;
+import com.microsoft.azure.servicebus.IMessageReceiver;
+import com.microsoft.azure.servicebus.primitives.ServiceBusException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Component;
+import uk.gov.hmcts.reform.slc.services.SendLetterJob;
+
+import java.util.function.Function;
+
+@Component
+public class MessageProcessor {
+    private static final Logger logger = LoggerFactory.getLogger(SendLetterJob.class);
+
+    private final IReceiverProvider receiverProvider;
+
+    public MessageProcessor(IReceiverProvider receiverProvider) {
+        this.receiverProvider = receiverProvider;
+    }
+
+    /**
+     * Reads message from a queue and send is as an argument to passed function.
+     */
+    public void handle(Function<IMessage, MessageHandlingResult> action) {
+
+        IMessageReceiver messageReceiver;
+
+        try {
+            messageReceiver = receiverProvider.get();
+        } catch (ConnectionException e) {
+            logger.error("Unable to connect to Service Bus", e);
+            return;
+        }
+
+        IMessage message = null;
+        try {
+            message = messageReceiver.receive();
+        } catch (InterruptedException | ServiceBusException e) {
+            logger.error("Unable to read message from queue", e);
+        }
+
+        if (message != null) {
+
+            MessageHandlingResult result = action.apply(message);
+            if (result == MessageHandlingResult.SUCCESS) {
+                try {
+                    messageReceiver.complete(message.getLockToken());
+                } catch (InterruptedException | ServiceBusException e) {
+                    logger.error("Unable to mark message " + message.getMessageId() + " as processed");
+                }
+            } else {
+                try {
+                    messageReceiver.deadLetter(message.getLockToken());
+                } catch (InterruptedException | ServiceBusException e) {
+                    logger.error("Unable to send message " + message.getMessageId() + " to deadletter subqueue", e);
+                }
+            }
+        } else {
+            logger.trace("No messages to process");
+        }
+
+        try {
+            messageReceiver.close();
+        } catch (ServiceBusException e) {
+            logger.error("Error closing connection");
+        }
+    }
```

Comment:
Hmm... When squinting eyes, a pattern emerges:

```java
try {
    messageReceiver.method(args);
} catch (InterruptedExecution | ServiceBusExecution e) {
    logger.error(message, e);
}
```

Should we leave it as is, or want to be uber-squeaky clean and refactor it?

Reply:
I'll consider adding some syntactic sugar / DSL on a future pull request.